### ISSUE TITLE
` "Never" does not derive from BaseException` false positive when raising `Never`

### ISSUE THREAD

**Issue by DetachHead** [microsoft/pyright#6619]
**Describe the bug**
since `Never` is a subtype of everything, it can be assigned to anything. however when raising a variable typed as `Never`, pyright thinks that it's not a valid subtype of `BaseException`

**Code or Screenshots**
```python
def main(foo: Never):
    _bar: BaseException = foo  # no error, because Never is a valid subtype of BaseException
    raise foo  # error: "Never" does not derive from BaseException
```
[playground](https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoByApgG5EgCwAUNQCZHBQQCGqAFMGGAFyGnkBKbtSiioAfQBGzELwBCzAM5EAogA8AxkQQwkYFFAC8UTmFEBiKCjPlwIADRRJRDcwCuyvmRCZFUZlAkzAA2SLRQim6S8AhEUGCMCsrqWjp6KCJiIKyephZQtri8AETE3sVQtGBEftb49CBIZCbg0Emqmtq6%2BkA)

**Comment by DetachHead** [comment]
> What does it mean to raise an error of type `Never`? Is there a practical use for what you're trying to do with this code?

i can't think of one but i still think it should be allowed, since the nature of `Never` means it's technically compatible with `BaseException`. i think it's more confusing to ban it imo, but i guess it doesn't matter too much

the reason i ran into this issue though was because of #6620, because i had code like this:
```py
with some_context_manager():
    try:
        result = fn(*args, **kwargs)
    except BaseException as e:
        error = e
        raise
if error:
    raise error # error was incorrectly narrowed to Never
```

**Comment by erictraut** [comment]
Yeah, I don't think there's any reason why this should matter, and I think the current behavior is justified, so I'm going to close it as "won't fix".

**Comment by mrcljx** [comment]
The problem appears when custom raising/assertion functions are involved and wanting to support other static analysis tools (that might not have full typing information). The following code is fine in both Pyright and Ruff (because the stdlib's `assert_never` is special-cased).

```py
from typing import assert_never

def get_type(value: str | int) -> str:
  match value:
    case str():
      return "str"
    case int():
      return "int"
    case _ as unreachable:
      assert_never(unreachable)
```

But when using a custom function `custom_assert_never(arg: Never) -> Never` and

```diff
- assert_never(unreachable)
+ custom_assert_never(unreachable)
```

then Ruff will trigger the following error because it doesn't have access to the type information:

```
RET503: Missing explicit `return` at the end of function able to return non-`None` value
```

With Mypy I was able to use add a `raise` here and make both tools happy (Mypy would detect the `raise` as unreachable and not complain about raising an unsupported type).

With Pyright I have to disable `RET503` on this line (disabling the rule globally works if the codebase is fully typed, but during the transition, it's great to have that rule catch inconsistent returns early).
